### DISCUSSION TITLE
Unhandled Rejection (SecurityError): Failed to execute 'texImage2D' on 'WebGL2RenderingContext': Tainted canvases may not be loaded.

### DISCUSSION THREAD

**Discussion by tusharsable** [iTwin/itwinjs-core#2747] [Q&A]
When we try to load image texture by sending a online URL. we get the following error
![image](https://user-images.githubusercontent.com/66006244/142570893-d2cc9f69-b395-4e8c-8504-de36d4f594bb.png)

When we use a local Image source we don't face the above error for the below code.
we apply the material to a plane and show in bentley viewer. 

we are currently using our app on localhost port and haven't deployed.

`async getMaterial(){
    
    const texture = await this.allocateTextureFromUrl(this.contourImage)
    var materialParams = new RenderMaterial.Params();
    var imodel = UiFramework.getIModelConnection()!
    materialParams.textureMapping = new TextureMapping(texture!, new TextureMapping.Params());
    var material = IModelApp.renderSystem.createMaterial(materialParams, imodel);
    return this.params.material = material; 
  
  }
  async allocateTextureFromUrl(url: string): Promise<RenderTexture | undefined> {

    const isOwned = true;
    const params = new RenderTexture.Params(undefined, undefined, isOwned);
    const textureImage = await imageElementFromUrl(url);
    return IModelApp.renderSystem.createTextureFromImage(textureImage, true, undefined, params);

  } `

We tried the solution mentioned [here ](https://stackoverflow.com/questions/30945945/tainted-canvases-may-not-be-loaded-cross-domain-issue-with-webgl-textures)for changing crossorigin to anonymous but didn't work for us.

**Comment by pmconne** [comment]
The client and the server [both have to agree](https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_enabled_image) to permit the use of cross-origin images.
If this image is served by you, configure your server to permit this; or consider bundling the image with your app.
